icu4c/source/common/lstmbe.cpp

```diff
@@ -0,0 +1,792 @@
+        return *this;
+    }
+
+    // Assign the values of another array of the same size into this one.
+    inline Array1D& assign(const ReadArray1D& a) {
+        U_ASSERT(a.d1() == d1());
+        for (int32_t i = 0; i < d1(); i++) {
+            data_[i] = a.get(i);
+        }
+        return *this;
+    }
+
+    // Apply tanh to all the elements in the array.
+    inline Array1D& tanh() {
+        for (int32_t i = 0; i < d1_; i++) {
+            data_[i] = std::tanh(data_[i]);
+        }
+        return *this;
+    }
+
+    // Apply sigmoid to all the elements in the array.
+    inline Array1D& sigmoid() {
+        for (int32_t i = 0; i < d1_; i++) {
+            data_[i] = 1.0/(1.0 + exp(-data_[i]));
+        }
+        return *this;
+    }
+
+    inline Array1D& clear() {
+        uprv_memset(data_, 0, d1_ * sizeof(float));
+        return *this;
+    }
+
+private:
+    void* memory_;
+    float* data_;
+    int32_t d1_;
+};
+
+Array1D::~Array1D()
+{
+    uprv_free(memory_);
+}
+
+class Array2D : public ReadArray2D {
+public:
+    Array2D() : memory_(nullptr), data_(nullptr), d1_(0), d2_(0) {}
+    Array2D(int32_t d1, int32_t d2)
+        : memory_(uprv_malloc(d1 * d2 * sizeof(float))),
+          data_((float*)memory_), d1_(d1), d2_(d2) {
+        clear();
+    }
+    virtual ~Array2D();
+
+    // ReadArray2D methods.
+    virtual int32_t d1() const { return d1_; }
+    virtual int32_t d2() const { return d2_; }
+    virtual float get(int32_t i, int32_t j) const {
+        U_ASSERT(i < d1_);
+        U_ASSERT(j < d2_);
+        return data_[i * d2_ + j];
+    }
+
+    inline Array1D row(int32_t i) const {
+        U_ASSERT(i < d1_);
+        return Array1D(data_ + i * d2_, d2_);
+    }
+
+    inline Array2D& clear() {
+        uprv_memset(data_, 0, d1_ * d2_ * sizeof(float));
+        return *this;
+    }
+
+private:
+    void* memory_;
+    float* data_;
+    int32_t d1_;
+    int32_t d2_;
+};
+
+Array2D::~Array2D()
+{
+    uprv_free(memory_);
+}
+
+typedef enum {
+    BEGIN,
+    INSIDE,
+    END,
+    SINGLE
+} LSTMClass;
+
+typedef enum {
+    UNKNOWN,
+    CODE_POINTS,
+    GRAPHEME_CLUSTER,
+} EmbeddingType;
+
+class LSTMData : public UMemory {
+public:
+    LSTMData();
+    virtual ~LSTMData();
+    virtual UHashtable* GetDictionary() const = 0;
+    virtual EmbeddingType type() const = 0;
+    virtual const UChar* name() const = 0;
+    virtual const ConstArray2D& embedding() const = 0;
+    virtual const ConstArray2D& forwardW() const = 0;
+    virtual const ConstArray2D& forwardU() const = 0;
+    virtual const ConstArray1D& forwardB() const = 0;
+    virtual const ConstArray2D& backwardW() const = 0;
+    virtual const ConstArray2D& backwardU() const = 0;
+    virtual const ConstArray1D& backwardB() const = 0;
+    virtual const ConstArray2D& outputW() const = 0;
+    virtual const ConstArray1D& outputB() const = 0;
+};
+
+LSTMData::LSTMData()
+{
+}
+
+LSTMData::~LSTMData()
+{
+}
+
+class LSTMResourceData : public LSTMData {
```

Comment:
Suggestion (optional): measure the performance difference of using a bare struct with fields, rather than a class with a vtable.  You only have 1 impl of the LSTMData interface.

Reply:
change to struct